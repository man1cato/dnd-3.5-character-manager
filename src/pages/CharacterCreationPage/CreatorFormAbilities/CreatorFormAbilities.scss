@import '../CharacterCreationPage';


.CreatorFormAbilities__grid {
   @extend .CreatorForm__grid;
   grid-template-columns: 2fr repeat(3, 1fr);
   margin-bottom: $m-size;
   input {
		width: 100%;
		height: 100%;
		text-align: center; 
   }
   button {
      padding: $s-size;
   }
}

.CreatorFormAbilities__error {
   @extend .CreatorForm__error;
   grid-column: 1 / span 4;
}