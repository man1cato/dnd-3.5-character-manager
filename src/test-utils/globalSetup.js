import fs from 'fs'
import getApi from '../utils/getApi'

export default async () => {
   const api = await getApi()
   
   fs.writeFile('src/test-utils/fixtures/api.json', JSON.stringify(api), (err) => {
      if (err) throw err
   })
}