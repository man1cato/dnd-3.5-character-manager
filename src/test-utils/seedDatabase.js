import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}