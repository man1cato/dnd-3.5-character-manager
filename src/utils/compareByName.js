export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}