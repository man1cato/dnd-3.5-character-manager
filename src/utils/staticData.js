export const schools = [
   'Abjuration','Clairsentience', 'Conjuration', 'Divination', 
   'Enchantment', 'Evocation', 'Illusion', 'Metacreativity', 
   'Necromancy', 'Psychokinesis', 'Psychometabolism', 'Psychoportation', 
   'Telepathy', 'Transmutation', 'Universal'
]

export const alignments = [
   'Lawful Good', 'Neutral Good', 'Chaotic Good', 
   'Lawful Neutral', 'True Neutral', 'Chaotic Neutral', 
   'Lawful Evil', 'Neutral Evil', 'Chaotic Evil'
]