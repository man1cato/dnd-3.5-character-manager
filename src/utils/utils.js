import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])