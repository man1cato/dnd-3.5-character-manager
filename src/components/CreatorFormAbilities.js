import React, { useEffect, Fragment } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'
import { rollDice, calcAbilityMod, calcSkillPoints } from '../utils/utils'
import { abilities } from '../utils/staticData'

const CreatorFormAbilities = ({
	values,
	selectedJobClass,
	selectedRace,
	setFieldValue, 
	validateForm
}) => {
	const raceMods = selectedRace.abilityMods

	const setSkillPoints = (intScore) => {
		let skillPoints = calcSkillPoints(selectedJobClass.name, calcAbilityMod(intScore)) * 4
		if (skillPoints < 1) { skillPoints = 1	} 
		if (selectedJobClass.name === 'Human') { skillPoints += 4}			
		setFieldValue('skillPoints', skillPoints, false)
		setFieldValue('remainingSkillPoints', skillPoints)
	}

	useEffect(() => {
		validateForm()
		if (!!values.abilities.int.final) { setSkillPoints(values.abilities.int.final) }
	}, [])

	return (
		<div>
			<h3 className="row--center">Ability Scores</h3>

			<div className="section form-grid--abilities">
				<h5 className="grid__col1 form-grid__header">Ability</h5>
				<h5 className="grid__col2 form-grid__header">Score</h5>
				<h5 className="grid__col3 form-grid__header">Race Mod</h5>
				<h5 className="grid__col4 form-grid__header">Final</h5>

				{_.keys(abilities).map((abbr) => {
					const fieldName = `abilities.${abbr}.score`
					const raceMod = raceMods[abbr]
					return (
						<Fragment key={abbr}>
							<div className="grid__col1">{abilities[abbr]} ({abbr.toUpperCase()})</div>
							<Field 
								className="grid__col2 number-input"
								type="number"
								name={fieldName}
								id={`${abbr}Input`}
								value={values.abilities[abbr].score}
								onChange={(e) => {
									const score = Number(e.target.value)
									const finalScore = score + raceMod
									console.log(score, finalScore)
									setFieldValue(`abilities.${abbr}.score`, score)
									setFieldValue(`abilities.${abbr}.final`, finalScore, false)
									if (abbr === 'int') { setSkillPoints(finalScore) }
								}}
							/>	
							<div className="grid__col3">{raceMod}</div>
							<div className="grid__col4">{values.abilities[abbr].final}</div>					
							<ErrorMessage className="grid__col-span4 form-group--error" name={fieldName} component="div" />
						</Fragment>
					)
				})}
			</div>

			<div className="row--center">
				<button
					className="button"
					id="abilitiesRollButton"
					type='button'
					onClick={() => {
						_.keys(abilities).forEach((abbr) => {
							const score = rollDice(6, 3)
							const finalScore = score + raceMods[abbr]
							setFieldValue(`abilities.${abbr}.score`, score)
							setFieldValue(`abilities.${abbr}.final`, finalScore, false)
							if (abbr === 'int') { setSkillPoints(finalScore) }
						})
					}}
				>
					Roll Scores
				</button>

			</div>
			
		</div>
	)
}

export default CreatorFormAbilities