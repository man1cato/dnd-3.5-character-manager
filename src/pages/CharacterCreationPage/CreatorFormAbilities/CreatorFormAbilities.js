import React, { useEffect, Fragment } from 'react'
import _ from 'lodash'
import { Field, ErrorMessage } from 'formik'

import { rollDice, calcAbilityMod, calcSkillPoints } from '../../../utils/utils'
import { abilities } from '../../../utils/staticData'
import './CreatorFormAbilities.scss'


const CreatorFormAbilities = ({
	values,
	selectedJobClass,
	selectedRace,
	setFieldValue, 
	validateForm
}) => {
	const raceMods = selectedRace.abilityMods

	const setSkillPoints = (intScore) => {
		let skillPoints = calcSkillPoints(selectedJobClass.name, calcAbilityMod(intScore)) * 4
		if (skillPoints < 1) { skillPoints = 1	} 
		if (selectedJobClass.name === 'Human') { skillPoints += 4}			
		setFieldValue('skillPoints', skillPoints, false)
		setFieldValue('remainingSkillPoints', skillPoints)
	}

	useEffect(() => {
		validateForm()
		if (!!values.abilities.int.final) { setSkillPoints(values.abilities.int.final) }
	}, [])

	return (
		<div className="container--body">					
			<h3 className="section__title">Ability Scores</h3>

			<div className="CreatorFormAbilities__grid">
				<h5 className="grid__col1">Ability</h5>
				<h5 className="grid__col2">Score</h5>
				<h5 className="grid__col3">Race Mod</h5>
				<h5 className="grid__col4">Final</h5>

				{_.keys(abilities).map((abbr) => {
					const fieldName = `abilities.${abbr}.score`
					const raceMod = raceMods[abbr]
					return (
						<Fragment key={abbr}>
							<div className="grid__col1">{abilities[abbr]} ({abbr.toUpperCase()})</div>
							<Field 
								className="grid__col2 number-input"
								type="number"
								name={fieldName}
								id={`${abbr}Input`}
								value={values.abilities[abbr].score}
								onChange={(e) => {
									const score = Number(e.target.value)
									const finalScore = score + raceMod
									setFieldValue(`abilities.${abbr}.score`, score)
									setFieldValue(`abilities.${abbr}.final`, finalScore, false)
									if (abbr === 'int') { setSkillPoints(finalScore) }
								}}
							/>	
							<div className="grid__col3">{raceMod}</div>
							<div className="grid__col4">{values.abilities[abbr].final}</div>					
							<ErrorMessage className="CreatorFormAbilities__error" name={fieldName} component="div" />
						</Fragment>
					)
				})}

				<button
					className="button grid__col2"
					id="abilitiesRollButton"
					type='button'
					onClick={() => {
						_.keys(abilities).forEach((abbr) => {
							const score = rollDice(3, 6)
							const finalScore = score + raceMods[abbr]
							setFieldValue(`abilities.${abbr}.score`, score)
							setFieldValue(`abilities.${abbr}.final`, finalScore, false)
							if (abbr === 'int') { setSkillPoints(finalScore) }
						})
					}}
				>
					Roll
				</button>
			</div>
			
		</div>
	)
}

export default CreatorFormAbilities