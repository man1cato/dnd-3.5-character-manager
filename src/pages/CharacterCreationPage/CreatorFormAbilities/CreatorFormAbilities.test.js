import React from 'react'
import { fireEvent } from '@testing-library/react'
import _ from 'lodash'

import CreatorFormAbilities from './CreatorFormAbilities'
import { abilities } from '../../../utils/staticData'
import { apiData, renderWithRedux, FormikWrapper } from '../../../test-utils/utils'
import { characterOne } from '../../../test-utils/seedDatabase'


const setFieldValue = jest.fn()
const validateForm = jest.fn()
let props
beforeAll(async () => {
   const api = await apiData()
   props = {
      values: { 
         race: api.races[characterOne.race],
         abilities: _.mapValues(abilities, () => ({ score: '' })) 
      },
      selectedJobClass: api.jobClasses[characterOne.jobClass],
      setFieldValue,
      validateForm
   }
})


test('should render CreatorFormAbilities correctly', () => {
   const { container } = renderWithRedux(<CreatorFormAbilities {...props} />, null, FormikWrapper)
   expect(container.firstChild).toMatchSnapshot()
})

test('should fill out all inputs with numbers when roll button is clicked', () => {
   // const { getByTestId } = render(<CreatorFormAbilities {...props} />)
   // fireEvent.click(getByTestId('abilitiesRollButton'))
   // expect(getByTestId('strInput').value).toEqual(expect.any(Number))
   // expect(getByTestId('dexInput').value).toEqual(expect.any(Number))
   // expect(getByTestId('conInput').value).toEqual(expect.any(Number))
   // expect(getByTestId('intInput').value).toEqual(expect.any(Number))
   // expect(getByTestId('wisInput').value).toEqual(expect.any(Number))
   // expect(getByTestId('chaInput').value).toEqual(expect.any(Number))
})