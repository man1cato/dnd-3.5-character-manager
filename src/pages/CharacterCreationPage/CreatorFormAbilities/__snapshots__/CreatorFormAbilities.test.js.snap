// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CreatorFormAbilities correctly 1`] = `
<div
  className="container--body"
>
  <h3
    className="section__title"
  >
    Ability Scores
  </h3>
  <div
    className="CreatorFormAbilities__grid"
  >
    <h5
      className="grid__col1"
    >
      Ability
    </h5>
    <h5
      className="grid__col2"
    >
      Score
    </h5>
    <h5
      className="grid__col3"
    >
      Race Mod
    </h5>
    <h5
      className="grid__col4"
    >
      Final
    </h5>
    <div
      className="grid__col1"
    >
      Strength
       (
      STR
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="strInput"
      name="abilities.str.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      -2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.str.score"
    />
    <div
      className="grid__col1"
    >
      Dexterity
       (
      DEX
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="dexInput"
      name="abilities.dex.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.dex.score"
    />
    <div
      className="grid__col1"
    >
      Constitution
       (
      CON
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="conInput"
      name="abilities.con.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      -2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.con.score"
    />
    <div
      className="grid__col1"
    >
      Intelligence
       (
      INT
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="intInput"
      name="abilities.int.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      2
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.int.score"
    />
    <div
      className="grid__col1"
    >
      Wisdom
       (
      WIS
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="wisInput"
      name="abilities.wis.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      0
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.wis.score"
    />
    <div
      className="grid__col1"
    >
      Charisma
       (
      CHA
      )
    </div>
    <FormikConnect(FieldInner)
      className="grid__col2 number-input"
      id="chaInput"
      name="abilities.cha.score"
      onChange={[Function]}
      type="number"
      value=""
    />
    <div
      className="grid__col3"
    >
      0
    </div>
    <div
      className="grid__col4"
    />
    <FormikConnect(ErrorMessageImpl)
      className="CreatorFormAbilities__error"
      component="div"
      name="abilities.cha.score"
    />
    <button
      className="button grid__col2"
      id="abilitiesRollButton"
      onClick={[Function]}
      type="button"
    >
      Roll
    </button>
  </div>
</div>
`;
